pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -892,6 +882,68 @@ cdef _timedelta_from_value_and_reso(int64_t value, NPY_DATETIMEUNIT reso):
     return td_base
 
 
+@cython.overflowcheck(True)
+cdef object create_timedelta(object value, str in_unit, NPY_DATETIMEUNIT out_reso):
+    """
+    Timedelta factory.
+
+    Overflow-safe, and allows for the creation of Timedeltas with non-nano resos while
+    the public API for that gets hashed out (ref: GH#46587). For now, Timedelta.__new__
+    just does arg validation and kwarg processing.
+
+    _timedelta_from_value_and_reso faster if value already an int that can be safely
+    cast to an int64.
+
+    Parameters
+    ----------
+    value : Timedelta, timedelta, np.timedelta64, str, int, float
+        The same value types accepted by Timedelta.__new__
+    in_unit : str
+        Denote the (np) unit of the input, if it's numeric
+    out_reso: NPY_DATETIMEUNIT
+        Desired resolution of new Timedelta
+
+    Notes
+    -----
+    Pass in_unit="ignore" (or "ns") with a numeric value to just do overflow checking
+    (and bypass the prior behavior of converting value -> td64[ns] -> int)
+    """
+    cdef:
+        int64_t out_value
+
+    if isinstance(value, _Timedelta):
+        return value
+
+    try:
+        # if unit == "ns", no need to create an m8[ns] just to read the (same) value back
+        # if unit == "ignore", assume caller wants to invoke an overflow-safe version of
+        # _timedelta_from_value_and_reso, and that any float rounding is acceptable
+        if (is_integer_object(value) or is_float_object(value)) and (
+            in_unit == "ns" or in_unit == "ignore"
+        ):
+            if util.is_nan(value):
+                return NaT
+            out_value = <int64_t>value
+        # is_timedelta_64_object may not give correct results w/ some versions? see e.g.
+        # github.com/pandas-dev/pandas/runs/6397652653?check_suite_focus=true#step:11:435
+        elif isinstance(value, np.timedelta64):
+            out_value = ensure_td64ns(value).view(np.int64)
+        elif isinstance(value, str):
+            if value.startswith(("P", "-P")):
+                out_value = parse_iso_format_string(value)
+            else:
+                out_value = parse_timedelta_string(value)
+        else:
+            out_value = convert_to_timedelta64(value, in_unit).view(np.int64)
+    except OverflowError as err:
```

Comment:
any way for less of this logic to be within the try/except?

Reply:
Agree, it's not ideal.

Initially, I tried to leave as much conditional processing as possible in `convert_to_timedelta64`, i.e. no 'if/else` here, but that led to considerable slowdowns in many benchmarks IIRC. So that would be a simpler, albeit slower, option.

A similar option would be to let all the dispatching on arg type happen automatically, and add overflow checking as needed within each variant, e.g. 

```python
@functools.singledispatch
def value_to_int(value):
    ...

@value_to_int.register(str)
def _(value: str):
    ...

def create_timedelta(value):
    ...
    try:
        out_value = value_to_int(value)
    catch OverflowError:
        raise OutOfBoundsTimedelta(msg) from err
    ...
```

I'm not sure if cython plays well with `singledispatch`? This approach would also expand the scope of this PR a lot.

Definitely open to other ideas though!